updating labels.md and design/labels.md

sdminonne opened the pull request:
@bgrant0607 @smarterclayton thanks to have a look

sdminonne committed:
updating labels.md and design/labels.md

bgrant0607 commented on docs/labels.md:
```diff
@@ -11,50 +10,66 @@ key/value labels set on it, with at most one label with a particular key.
+Set based requirements can be mixed with equality based requirements. For example: `partition,environment!=qa`.
+
 
 LIST and WATCH operations may specify label selectors to filter the sets of objects returned using a query parameter: `?labels=key1%3Dvalue1,key2%3Dvalue2,...`.
```
Please extend this example to illustrate the expanded query syntax.

bgrant0607 commented on docs/design/labels.md:
```diff
@@ -10,42 +10,15 @@ Each object can have a set of key/value labels set on it, with at most one label
-
-Label selectors may also be used to associate policies with sets of objects.
-
 We also [plan](https://github.com/GoogleCloudPlatform/kubernetes/issues/560) to make labels available inside pods and [lifecycle hooks](/docs/container-environment.md).
```
Please delete this line for now.

bgrant0607 commented on docs/labels.md:
```diff
@@ -1,9 +1,8 @@
-created by users at the same time as an object.  Each object can have a set of
-key/value labels set on it, with at most one label with a particular key. 
+Labels can be used to organize and to select subsets of objects.  Labels can be attached to objects at creation time and subsequently added and modified at any time.
+Each object can have a set of key/value labels defined.  Each Key is unique for a given object.
```
Each Key must be unique for a given object.

bgrant0607 commented on docs/labels.md:
```diff
@@ -11,50 +10,66 @@ key/value labels set on it, with at most one label with a particular key.
+Unlike [names and UIDs](identifiers.md), labels do not provide uniqueness. In general, we expect many objects to carry the same label(s).
+Sets identified by labels could be overlapping (think Venn diagrams). For instance, a service might target all pods with `"tier": "frontend"` and  `"environment" : "prod"`.  Now say you have 10 replicated pods that make up this tier.  But you want to be able to 'canary' a new version of this component.  You could set up a `replicationController` (with `replicas` set to 9) for the bulk of the repl ...
+
+Valid label keys are comprised of two segments - prefix and name - separated by a slash (`/`).  The name segment is required and must be a DNS label: 63 characters or less, all lowercase, beginning and ending with an alphanumeric character (`[a-z0-9]`), with dashes (`-`) and alphanumerics between.  The prefix and slash are optional.  If specified, the prefix must be a DNS
```
Please move this up above "Unlike names and UIDs..."

bgrant0607 replied:
Also, we extended the syntax to allow uppercase and underscores:
https://github.com/GoogleCloudPlatform/kubernetes/blob/master/pkg/util/validation.go#L46

Label values must be no more than 63 characters and are restricted to the same character set, but may be the empty string:
https://github.com/GoogleCloudPlatform/kubernetes/blob/master/pkg/util/validation.go#L33

bgrant0607 replied:
Ah, I see you cover label values below.

bgrant0607 commented on docs/labels.md:
````diff
@@ -11,50 +10,66 @@ key/value labels set on it, with at most one label with a particular key.
-```
-DNS_LABEL and DNS_SUBDOMAIN are defined in the [identifiers design doc](/docs/design/identifiers.md). The prefix is optional. If the prefix is not specified, the key is assumed to be private to the user. Other system components that wish to use labels must specify a prefix. The "kubernetes.io/" prefix is reserved for use by kubernetes components.
+Unlike [names and UIDs](identifiers.md), labels do not provide uniqueness. In general, we expect many objects to carry the same label(s).
+Sets identified by labels could be overlapping (think Venn diagrams). For instance, a service might target all pods with `"tier": "frontend"` and  `"environment" : "prod"`.  Now say you have 10 replicated pods that make up this tier.  But you want to be able to 'canary' a new version of this component.  You could set up a `replicationController` (with `replicas` set to 9) for the bulk of the repl ...
````
We should move this paragraph down to the label selector section, which goes into this in more detail.

bgrant0607 commented on docs/labels.md:
```diff
@@ -11,50 +10,66 @@ key/value labels set on it, with at most one label with a particular key.
+## Label selectors
+
+Via a label selector, the client/user can identify a set of objects. The label selector is the core grouping primitive in Kubernetes.
+The API currently supports two types of selectors: equality based and set based.
```
equality-based and set-based (need hyphens)

bgrant0607 commented on docs/labels.md:
```diff
@@ -11,50 +10,66 @@ key/value labels set on it, with at most one label with a particular key.
+
+Via a label selector, the client/user can identify a set of objects. The label selector is the core grouping primitive in Kubernetes.
+The API currently supports two types of selectors: equality based and set based.
+A label selector can be made of multiple _requirements_ which are comma separated. In case of multiple requirements, all must be satisfied so comma separaator acts as an AND logical operator.
```
comma-separated

bgrant0607 replied:
In *the* case of multiple...

bgrant0607 replied:
typo: separator

bgrant0607 commented on docs/labels.md:
```diff
@@ -11,50 +10,66 @@ key/value labels set on it, with at most one label with a particular key.
+
+### Equality based requirement
+
+Equality (or inequality) based requirements allow filtering by label keys and values. Matching objects must have all of the specified labels (both keys and values), though they may have additional labels as well.
```
Equality- or inequality-based

bgrant0607 commented on docs/labels.md:
```diff
@@ -11,50 +10,66 @@ key/value labels set on it, with at most one label with a particular key.
+A label selector can be made of multiple _requirements_ which are comma separated. In case of multiple requirements, all must be satisfied so comma separaator acts as an AND logical operator.
+
+
+### Equality based requirement
```
Equality-based

bgrant0607 commented on docs/labels.md:
```diff
@@ -11,50 +10,66 @@ key/value labels set on it, with at most one label with a particular key.
+The latter selects all resources with key equal to `tier` and value distinct from `frontend`.
+One could filter for resources in `production` but not `frontend` using the comma operator: `environment=production,tier!=frontend`
+
+### Set based requirement
```
Set-based

bgrant0607 replied:
Just fix the hyphenation everywhere.

bgrant0607 commented on docs/labels.md:
```diff
@@ -11,50 +10,66 @@ key/value labels set on it, with at most one label with a particular key.
 
 LIST and WATCH operations may specify label selectors to filter the sets of objects returned using a query parameter: `?labels=key1%3Dvalue1,key2%3Dvalue2,...`.
 
-The `service` and `replicationController` kinds of objects use selectors to match sets of pods that they operate on.
```
Somewhere here we need a short section on selectors in API objects, such as service and replication controller, with an example using the map syntax and clarifying that only equality-based selectors are currently supported.

bgrant0607 commented on docs/design/labels.md:
```diff
@@ -53,23 +26,15 @@ Kubernetes also currently supports two objects that use label selectors to keep
 
-Sets identified by labels and label selectors could be overlapping (think Venn diagrams). For instance, a service might target all pods with `tier in (frontend), environment in (prod)`.  Now say you have 10 replicated pods that make up this tier.  But you want to be able to 'canary' a new version of this component.  You could set up a `replicationController` (with `replicas` set to 9) for the bul ...
-
-Note that the superset described in the previous example is also heterogeneous. In long-lived, highly available, horizontally scaled, distributed, continuously evolving service applications, heterogeneity is inevitable, due to canaries, incremental rollouts, live reconfiguration, simultaneous updates and auto-scaling, hardware upgrades, and so on.
```
Did you not like this sentence? It looks like it was just removed, not moved.

sdminonne replied:
I did but I find (this sentence good for an introduction for label users). I didn't find this sentence appropriate in the `design` side. I'll reintegrate it.

bgrant0607 commented:
Thanks @sdminonne. 

The 2 labels docs were originally a single doc. There's still a lot of overlap. What do you think about re-merging them? There's really just one sentence in the design doc that's implementation-oriented rather than usage-oriented.

sdminonne commented:
@bgrant0607 working on this. Need few hours to come back.

bgrant0607 commented:
Thanks.

sdminonne commented:
@bgrant0607  PTAL
- removed doc/design/labels.md (and tried to merge in docs/labels.md)
- hyphen requirements
- removed reference to https://github.com/GoogleCloudPlatform/kubernetes/issues/560
- added a more formal section for  charset (not requested in reality) but I think it's better
- added URL for set-based. Not sure I've understood the map based example https://github.com/GoogleCloudPlatform/kubernetes/pull/5189#discussion_r26142826 ... I'm OK to add one but can you re-explain this, sorry for this.
 
Hope it's better now.

bgrant0607 commented on docs/labels.md:
```diff
@@ -1,9 +1,8 @@
 # Labels
 
 _Labels_ are key/value pairs that are attached to objects, such as pods.
```
Please add after this sentence:

Labels are intended to be used to specify identifying attributes of objects that are meaningful and relevant to users, but which do not directly imply semantics to the core system.

bgrant0607 commented on docs/labels.md:
```diff
@@ -11,50 +10,93 @@ key/value labels set on it, with at most one label with a particular key.
+## Motivation
 
-Unlike [names and UIDs](identifiers.md), labels do not provide uniqueness. In general, we expect many objects to carry the same label(s). 
+Service deployments and batch processing pipelines are often multi-dimensional entities (e.g., multiple partitions or deployments, multiple release tracks, multiple tiers, multiple micro-services per tier). Management often requires cross-cutting operations, which breaks encapsulation of strictly hierarchical representations, especially rigid hierarchies determined by the infrastructure rather th ...
```
Please move the last sentence (Labels enable users...) first, and put it in a paragraph on its own.

bgrant0607 commented on docs/labels.md:
```diff
@@ -11,50 +10,93 @@ key/value labels set on it, with at most one label with a particular key.
-
-We also [plan](https://github.com/GoogleCloudPlatform/kubernetes/issues/560) to make labels available inside pods and [lifecycle hooks](container-environment.md).
 
 Labels let you categorize objects in a complex service deployment or batch processing pipelines along multiple
```
This sentence is fairly redundant with the above paragraph. I would shorten it to just "Example labels:"

bgrant0607 commented on docs/labels.md:
```diff
@@ -11,50 +10,93 @@ key/value labels set on it, with at most one label with a particular key.
+Pods (and other objects) may belong to multiple sets simultaneously, which enables representation of service substructure and/or superstructure. In particular, labels are intended to facilitate the creation of non-hierarchical, multi-dimensional deployment structures. They are useful for a variety of management purposes (e.g., configuration, deployment) and for application introspection and analy ...
+
+
+Via a label selector, the client/user can identify a set of objects. The label selector is the core grouping primitive in Kubernetes.
```
This should be the first sentence in this subsection. "label selector" should be italicized.

bgrant0607 commented on docs/labels.md:
```diff
@@ -11,50 +10,93 @@ key/value labels set on it, with at most one label with a particular key.
+
+Unlike [names and UIDs](identifiers.md), labels do not provide uniqueness. In general, we expect many objects to carry the same label(s).
+
+Sets identified by labels could be overlapping (think Venn diagrams). For instance, a service might target all pods with `"tier": "frontend"` and  `"environment" : "prod"`.  Now say you have 10 replicated pods that make up this tier.  But you want to be able to 'canary' a new version of this component.  You could set up a `replicationController` (with `replicas` set to 9) for the bulk of the repl ...
```
I'd move these 3 paragraphs to just above "Future developments", below the discussion of the use of label selectors in services and replication controllers.

bgrant0607 commented on docs/labels.md:
```diff
@@ -11,50 +10,93 @@ key/value labels set on it, with at most one label with a particular key.
+Concerning API: we may extend such filtering to DELETE operations in the future.
 
-The `service` and `replicationController` kinds of objects use selectors to match sets of pods that they operate on.
+We'll eventually index and reverse-index labels for efficient queries and watches, use them to sort and group in UIs and CLIs, etc. We don't want to pollute labels with non-identifying, especially large and/or structured, data. Non-identifying information should be recorded using [annotations](/docs/annotations.md).
```
Please put the latter 2 sentences about annotations to a paragraph on their own, just above the Motivation section.

bgrant0607 commented on docs/labels.md:
```diff
@@ -11,50 +10,93 @@ key/value labels set on it, with at most one label with a particular key.
+We'll eventually index and reverse-index labels for efficient queries and watches, use them to sort and group in UIs and CLIs, etc. We don't want to pollute labels with non-identifying, especially large and/or structured, data. Non-identifying information should be recorded using [annotations](/docs/annotations.md).
 
-See the [Labels Design Document](./design/labels.md) for more about how we expect labels and selectors to be used, and planned features.
+In the future, label selectors will be used to identify other types of distributed service workers, such as worker pool members or peers in a distributed application.
```
Please just delete this sentence. It's more about services than labels.

bgrant0607 commented:
Thanks a lot. I made a few more suggestions but I think this will be good once they're done. After you make the changes, please squash your commits and comment on the PR.

sdminonne commented:
@bgrant0607, squashed and rebased. PTAL.

bgrant0607 commented:
Perfect!
